### PR TITLE
os/newstrore:Implement fiemap

### PR DESCRIPTION
**Pull request by xiaoxichen**
For simplicity we ignore holes inside an fragment now.

Signed-off-by: Xiaoxi Chen <xiaoxi.chen@intel.com>

### CODE REVIEW

**Comment by dachary** [comment]
@xiaoxichen please ignore the bot failure: master is currently failing the make check tests.

File: src/os/newstore/NewStore.cc [hunk 1]
```diff
@@ -1368,7 +1367,95 @@ int NewStore::fiemap(
...
+	++fp;
+      continue;
+    }
+    // we are seeing a hole, time to add an entry to fiemap.
+    m[start] = offset - start;
+    dout(20) << __func__ << " get fiemap entry, off =  " << start << " len=" << m[start] << dendl;
+    offset += x_len;
+    start = offset;
+    len -= x_len;
+    continue;
```
Comment (liewegas) [inline src/os/newstore/NewStore.cc]: I think this misses the end case.. if the whole object is allocated, we need an entry in the map.  Or do the callers assume an empty map means fully allocated?

Comment (xiaoxichen) [inline src/os/newstore/NewStore.cc]: hmm....sorry....that is a bug....

**Comment by xiaoxichen** [comment]
@liewegas , fixed & updated

Hard to write an unit test, do that in a hack way,see below cases, all passed.

  o->onode.size = 4096;
  
  o->onode.data_map[0] = fragment_t(0, 2048);      //[0,2048)
  o->onode.data_map[3000] = fragment_t(0, 1000);    //[3000,4000)
  
  o->onode.overlay_map[0] = overlay_t(0, 0, 64);    //[0,64)
  o->onode.overlay_map[64] = overlay_t(1, 0, 64);   //[64,128)
  o->onode.overlay_map[2000] = overlay_t(2, 0, 64); //[2000~2064) front overlap with fragment0 
  o->onode.overlay_map[2800] = overlay_t(3, 0, 64); //[2800~2864) standalone
  o->onode.overlay_map[2950] = overlay_t(4, 0 , 64); // [2950~3014) tail overlap with fragment 1
  

    fiemap 0~2048 size 4096
    fiemap get overlay, off =  0 len=64
    fiemap get overlay, off =  64 len=64
    fiemap get frag, off =  128 len=1872
    fiemap get overlay, off =  2000 len=48
    fiemap get fiemap entry, off =  0 len=2048
    fiemap final fiemap, off = 0 len = 2048


    fiemap 0~1024 size 4096
    fiemap get overlay, off =  0 len=64
    fiemap get overlay, off =  64 len=64
    fiemap get frag, off =  128 len=896
    fiemap get fiemap entry, off =  0 len=1024
    fiemap final fiemap, off = 0 len = 1024


    fiemap 0~96 size 4096
    fiemap get overlay, off =  0 len=64
    fiemap get overlay, off =  64 len=32
    fiemap get fiemap entry, off =  0 len=96
    fiemap final fiemap, off = 0 len = 96


    fiemap 0~2500 size 4096
    fiemap get overlay, off =  0 len=64
    fiemap get overlay, off =  64 len=64
    fiemap get frag, off =  128 len=1872
    fiemap get overlay, off =  2000 len=64
    fiemap get fiemap entry, off =  0 len=2064
    fiemap final fiemap, off = 0 len = 2064


    fiemap 0~2600 size 4096
    fiemap get overlay, off =  0 len=64
    fiemap get overlay, off =  64 len=64
    fiemap get frag, off =  128 len=1872
    fiemap get overlay, off =  2000 len=64
    fiemap get fiemap entry, off =  0 len=2064
    fiemap final fiemap, off = 0 len = 2064


    fiemap 0~3300 size 4096
    fiemap get overlay, off =  0 len=64
    fiemap get overlay, off =  64 len=64
    fiemap get frag, off =  128 len=1872
    fiemap get overlay, off =  2000 len=64
    fiemap get fiemap entry, off =  0 len=2064
    fiemap get overlay, off =  2800 len=64
    fiemap get fiemap entry, off =  2800 len=64
    fiemap get overlay, off =  2950 len=64
    fiemap get frag, off =  3014 len=286
    fiemap get fiemap entry, off =  2950 len=350
    fiemap final fiemap, off = 0 len = 2064
    fiemap final fiemap, off = 2800 len = 64
    fiemap final fiemap, off = 2950 len = 350


    fiemap 0~4096 size 4096
    fiemap get overlay, off =  0 len=64
    fiemap get overlay, off =  64 len=64
    fiemap get frag, off =  128 len=1872
    fiemap get overlay, off =  2000 len=64
    fiemap get fiemap entry, off =  0 len=2064
    fiemap get overlay, off =  2800 len=64
    fiemap get fiemap entry, off =  2800 len=64
    fiemap get overlay, off =  2950 len=64
    fiemap get frag, off =  3014 len=986
 .266129   fiemap get fiemap entry, off =  2950 len=1050
 .266130   fiemap final fiemap, off = 0 len = 2064
 .266130   fiemap final fiemap, off = 2800 len = 64
 .266131   fiemap final fiemap, off = 2950 len = 1050

**Comment by xiaoxichen** [comment]
@liewegas ping?

**Comment by xiaoxichen** [comment]
@liewegas updated.

**Pull request merged**